File: programming_questions/arrays_based/Max_Profit_on_Subset.go
package main

import "fmt"

func main() {
	var inputArray []int = []int{3, 6, 36, 32, 32, 121, 66, 24, 26, 371, 661, 6, 4, 8}
	//var input int = 0
	var lastProfit int = 0

	/*for true {
		fmt.Scanf("%d", &input)
		if input == -1 {
			break
		}
		inputArray = append(inputArray, input)
	}*/

	size := len(inputArray)
	if size < 2 {
		return
	}

	for i, j := 0, 0; i < size; {
		for j = i; j+1 < size && inputArray[j] <= inputArray[j+1]; j++ {
		}
		if i != j {
			currentProfit := inputArray[j] - inputArray[i]
			if currentProfit > lastProfit {
				lastProfit = currentProfit
			}
		}
		i = j + 1
	}

	//fmt.Println(size)
	//fmt.Println(inputArray)
	fmt.Println(lastProfit)
}

File: programming_questions/arrays_based/Max_Profit_on_Subset.c
#include <stdio.h>
#include <stdlib.h>

#define ARRAY_SIZE 14

int inputArray[ARRAY_SIZE] = {3,6,36,32,32,121,66,24,26,371,661,6,4,8};
int size = ARRAY_SIZE;

int main() {
  //int inputArray[ARRAY_SIZE], size = 0;
  int currentProfit,lastProfit=0;

  /*do {
    scanf("%d", &inputArray[size++]);
    } while (inputArray[size-1] != -1);*/

  //size--;

  if (size < 2) {
    printf("%d",0);
    return -1;
  }

  for (int i=0,j=0; i<size; i=j+1) {
    for (j = i; j+1<size && inputArray[j] <= inputArray[j+1]; j++);
    if (j != i) {
      currentProfit = inputArray[j]-inputArray[i];
      if (lastProfit < currentProfit)
        lastProfit = currentProfit;
    }
  }

  printf("%d", lastProfit);
  return 0;
}
